File: Examples/003/Fort/Solution/pi_cal.f90
program pi_calc
    use mpi
    implicit none
    integer :: rank, size, ierr
    integer(kind=8) :: points_per_rank = 500000000
    integer(kind=8) :: local_inside, global_inside
    real(kind=8) :: pi_estimate

    call MPI_Init(ierr)
    call MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)
    call MPI_Comm_size(MPI_COMM_WORLD, size, ierr)

    ! Encapsulated call for profiler visibility
    local_inside = run_monte_carlo_vector(points_per_rank)

    call MPI_Reduce(local_inside, global_inside, 1, MPI_INTEGER8, MPI_SUM, 0, MPI_COMM_WORLD, ierr)

    if (rank == 0) then
        pi_estimate = 4.0d0 * real(global_inside, 8) / real(points_per_rank * size, 8)
        print *, "Estimated Pi (Vectorized): ", pi_estimate
    end if

    call MPI_Finalize(ierr)

contains

    integer(kind=8) function run_monte_carlo_vector(n)
        integer(kind=8), intent(in) :: n
        integer(kind=8) :: i, chunk
        integer, parameter :: CHUNK_SIZE = 10000
        real(kind=8) :: x(CHUNK_SIZE), y(CHUNK_SIZE)
        
        run_monte_carlo_vector = 0
        do chunk = 1, n, CHUNK_SIZE
            ! Fortran Intrinsic Vectorization
            call random_number(x)
            call random_number(y)
            do i = 1, CHUNK_SIZE
                if (x(i)*x(i) + y(i)*y(i) <= 1.0d0) then
                    run_monte_carlo_vector = run_monte_carlo_vector + 1
                end if
            end do
        end do
    end function run_monte_carlo_vector

end program pi_calc

File: Examples/004/CPP/Problem/pi_cal.cpp
#include <iostream>
#include <mpi.h>
#include <cstdlib>

// Encapsulated for visibility in the profiler
long long run_monte_carlo(long long n) {
    long long inside = 0;
    for (long long i = 0; i < n; ++i) {
        double x = (double)rand() / RAND_MAX;
        double y = (double)rand() / RAND_MAX;
        if (x * x + y * y <= 1.0) {
            inside++;
        }
    }
    return inside;
}

int main(int argc, char** argv) {
    MPI_Init(&argc, &argv);
    int rank, size;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    long long points_per_rank = 500000000;
    srand(rank * 12345); 

    // Calculation is now its own region in Score-P
    long long local_inside = run_monte_carlo(points_per_rank);

    long long global_inside = 0;
    MPI_Reduce(&local_inside, &global_inside, 1, MPI_LONG_LONG, MPI_SUM, 0, MPI_COMM_WORLD);

    if (rank == 0) {
        double pi = 4.0 * (double)global_inside / (points_per_rank * size);
        std::cout << "Estimated Pi (Slow Rand): " << pi << std::endl;
    }

    MPI_Finalize();
    return 0;
}
